language-server/src/features/document-settings.js

```diff
@@ -26,6 +39,7 @@ export default {
       } else {
         globalSettings = change.settings.jsonSchemaLanguageServer ?? globalSettings;
       }
+      schemaFilePatterns = globalSettings.schemaFilePatterns ?? ["**/*.schema.json", "**/schema.json"];
```

Comment:
This appears hardcoded in two places (also line 66). It should be extracted to a variable.

```javascript
["**/*.schema.json", "**/schema.json"]
```

Reply:
Right, replaced it with `defaultFilePatterns` for now. or are you expecting this to be in `globalSettings` too?

Reply:
@jdesrosiers I've hardcoded the value for now. should I wrap this up in `globalSettings` and instead of exporting `schemafilePatterns` I export `globalSettings` and use `globalSettings.schemaFilePatterns`?

Reply:
I've extracted it into a variable for now. should I wrap this up in `globalSettings` and instead of exporting `schemafilePatterns` I export `globalSettings` and use `globalSettings.schemaFilePatterns`?

Reply:
> I've extracted it into a variable for now. should I wrap this up in `globalSettings`

What do you think? I keep forgetting the purpose of `globalSettings`. I think its purpose is to be the default settings for when the client doesn't support `getConfiguration` or the result of that call is null/undefined. In that case, it makes sense for `schemaFilePatterns` to be in `globalSettings`.

But `globalSettings` also gets set when `onDidChangeConfiguration` is received. So, if the client sends configuration without `schemaFilePatterns`, then everything would break. So, it seems we need `defaultSchemaFilePatterns` so it can't get overwritten and exists when it's needed.

But, the part I'm not sure of is, what's the reason `onDidChangeConfiguration` overwrites `globalSettings`? I guess this could be useful if the client sends `onDidChangeConfiguration` requests, but doesn't support `getConfiguration` requests. That doesn't make sense to me. So, if overwriting `globalSettings` isn't the right thing, then putting the default `schemFilePatterns` in `globalSettings` makes sense.

I really don't know what the right thing is here. What do you think?

Reply:
> instead of exporting `schemafilePatterns` I export `globalSettings` and use `globalSettings.schemaFilePatterns`?

I don't think you should be exporting any of those. I don't think `schemaFilePatterns` is even necessary. When we need `schemaFilePatterns`, we should be calling `getDocumentSettings` and getting the value from there.

Reply:
> But, the part I'm not sure of is, what's the reason `onDidChangeConfiguration` overwrites `globalSettings`? I guess this could be useful if the client sends `onDidChangeConfiguration` requests, but doesn't support `getConfiguration` requests. That doesn't make sense to me. So, if overwriting `globalSettings` isn't the right thing, then putting the default `schemFilePatterns` in `globalSettings` makes sense.

Right, my understanding of  `globalSettings` aligns the same way, this is how it is used in the documentation example and that is the same analogy used in our extension for now not fully knowing where things would fail.  I'm not sure if the comments here mean multiple clients but I think that would still not justify a client sending `onDidChangeConfiguration` requests but doesn't support `getConfiguration`. 

```
// The global settings, used when the `workspace/configuration` request is not supported by the client.
// Please note that this is not the case when using this server with the client provided in this example
// but could happen with other clients.
const defaultSettings: ExampleSettings = { maxNumberOfProblems: 1000 };
let globalSettings: ExampleSettings = defaultSettings;

// Cache the settings of all open documents
let documentSettings: Map<string, Thenable<ExampleSettings>> = new Map();

connection.onDidChangeConfiguration(change => {
  if (hasConfigurationCapability) {
    // Reset all cached document settings
    documentSettings.clear();
  } else {
    globalSettings = <ExampleSettings>(
      (change.settings.languageServerExample || defaultSettings)
    );
  }
```

Reply:
Agreed. Overwriting `globalSettings` doesn't seem to make sense. There might be some client out there where it does, but we don't understand that situation. So, for now, let's just code for those clients we're targeting (vscode and neovim). I'd rather not have code for situations we don't understand.

Let's drop the idea of `globalSettings` and just do something to handle defaults for settings that are required like `schemaFilePatterns`.

Reply:
> Let's drop the idea of `globalSettings` and just do something to handle defaults for settings that are required like `schemaFilePatterns`.

I've handled it in `defaultSettings`  and changed the `onDidChangeConfiguration`  request but doesn't support configuration capability updating the `globalSettings`.